### ISSUE TITLE
brod_group_subscriber_v2 fails to resubscribe on brod_client crash

### ISSUE THREAD

**Issue by epsylonix** [kafka4beam/brod#413]
Hello!

I believe there is a bug in `brod_group_subscriber_v2` that could leave it permanently unsubscribed.

`brod_client` might terminate for various reasons. In that case `brod_topic_subscriber` [exits](https://github.com/klarna/brod/blob/master/src/brod_topic_subscriber.erl#L349-L352) assuming it will be restarted but `brod_group_subscriber_v2` [doesn't crash with it]( https://github.com/klarna/brod/blob/master/src/brod_group_subscriber_v2.erl#L494) and [ignores its `DOWN` message](https://github.com/klarna/brod/blob/master/src/brod_group_subscriber_v2.erl#L366). `brod_client` is restarted by its supervisor but `brod_topic_subscriber` is not. That leaves the consumer unsubscribed and it will not be live again untill the whole application is restarted manually.

This can be reproduced locally by restarting the Kafka broker:
- `brod_client` crashes
- `brod_topic_subscriber` exits
- `brod_group_subscriber_v2` logs "Received DOWN message from unknown process"
- `brod_client` gets restarted but `brod_group_subscriber_v2` will not receive Kafka messages anymore.

Could you please elaborate on the issue of `brod_group_subscriber_v2` stability?  There is no mention in the documentation if it is to be considered experimental or not safe for production use but going through the list of recent commits to the module I got the feeling that it might not be production-ready yet. Is it used in production anywhere?

Thanks!

**Comment by k32** [comment]
Hi! 

> brod_group_subscriber_v2 doesn't crash with it 

In the above code v2 behavior unlinks from the worker process rather than the client. It still monitors the workers and will handle worker failure in `handle_info`. Failure to communicate with the client should propagate to the `group_coordinator` and the partition worker processes, and bring down the group_subscriber_v2.

> That leaves the consumer unsubscribed and it will not be live again untill the whole application is restarted manually.

Did you actually observe this bug somewhere? It would be interesting to look at a minimal example that triggers the problem, because I couldn't reproduce it in this small testcase:

```erlang
t_client_crash(Config) when is_list(Config) ->
  process_flag(trap_exit, true),
  _ = produce_payloads(?topic, 0, Config),
  {ok, SubscriberPid} = start_subscriber(Config, [?topic], #{}),
  timer:sleep(5000),
  brod:stop_client(?CLIENT_ID),
  receive
    {'EXIT', SubscriberPid, _} ->
      ok
  after 10000 ->
      error(subscriber_did_not_die)
  end.
```

edit: I was thinking that maybe a subscriber that didn't get any partition assignment could demonstrate the described behavior, but even this more complicated test passes:

```erlang
t_client_crash(Config) when is_list(Config) ->
  process_flag(trap_exit, true),
  _ = produce_payloads(?topic, 0, Config),
  {ok, SubscriberPid1} = start_subscriber(Config, [?topic], #{}),
  {ok, SubscriberPid2} = start_subscriber(Config, [?topic], #{}),
  timer:sleep(5000),
  brod:stop_client(?CLIENT_ID),
  receive
    {'EXIT', SubscriberPid1, _} ->
        receive
          {'EXIT', SubscriberPid2, _} ->
            ok
        after 10000 ->
            error(subscriber2_did_not_die)
        end
  after 10000 ->
      error(subscriber1_did_not_die)
  end.
```

edit2: I was answering to an older version of the issue...

**Comment by epsylonix** [comment]
It seems my understanding of what happens when `brod_client` crashes was flawed :unamused: 

I was investigating a problem when the consumer somehow didn't receive assignments after the `brod_client` crash. I expected that after `brod_client` crashes, `topic_subscriber` [would exit](https://github.com/klarna/brod/blob/master/src/brod_topic_subscriber.erl#L347-L352) causing `brod_group_subscriber_v2` to exit too, and I didn't find the logged crash for the subscriber process but there was a bunch of `Received DOWN message from unknown process` logs so I assumed the wrong `DOWN` message was ignored. 

As far as I can understand what actually happens when a client crashes is `brod_group_subscriber_v2` receives `unsubscribe_all_partitions` message before the topic subscribers exit, at which point it terminates all the topic subscribers (workers) and removes their pids from the state. But as it still monitors them it receives their `DOWN` messages anyway and logs them as `DOWN message from unknown process` (maybe workers should be `demonitor`ed before terminated).

I checked again the steps to reproduce the problem that I described above and it doesn't work indeed. While there were no crashes logged for `brod_group_subscriber_v2`, it seems that I had mistaken the crash of the linked `brod_group_coordinator` process for the `brod_client` process crash and didn't actually verify if `brod_group_subscriber_v2` process was alive at that point. 

Sorry I mislead you with the description above and wasted your time. I'll try to identify the cause of the problem I investigated initially and if it turns out to be brod's fault I'll create another issue. Thanks!
